File: Data Structures & Algorithms/maximum-subarray/submission-5.cpp
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int n = nums.size();
        int cur = nums[0], res = nums[0];

        for(int i = 1; i < n; i++){
            cur = max(nums[i], cur + nums[i]);
            res = max(res, cur);
        }

        return res;
    }
};

File: Data Structures & Algorithms/coin-change/submission-5.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        cache = {}
        
        def dfs(cur):
            if cur == amount:
                return 0

            if cur in cache:
                return cache[cur]

            minCoin = float('inf')
            for coin in coins:
                if cur + coin <= amount:
                    minCoin = min(minCoin, 1 + dfs(cur + coin))
            
            cache[cur] = minCoin
            return minCoin

        res = dfs(0)

        return -1 if res == float('inf') else res
